Add `--no-newline` flag

Qix- opened the pull request:
Closes #13.

Opening a PR because I'd rather have a single `-n` option instead of also allowing `--bare`, but not sure if Meow can do that.

Qix- committed:
add -n option (closes #13)

sindresorhus commented on cli.js:
```diff
@@ -55,6 +56,10 @@ const cli = meow(`
 		},
+		bare: {
+			type: 'boolean',
+			alias: 'n'
```
Why `-n`? Shouldn't it be `-b`?

And I don't really think it's common enough of a use-case to warrant an alias.

Qix- replied:
It's the same flag as `echo`.

msabramo replied:
Since the main idea appears to be to support `-n` like echo, which I like, and the `--bare` seemed really only to make meow happy, I’d like to suggest changing `--bare` to `--nonewline` or `--no-newline`, which I think is more descriptive and maybe resolves the question about non-matchiness of short and long flags from @sindresorhus better.

sindresorhus replied:
> --no-newline

👍

Qix- replied:
`meow` doesn't support this, at least I can't figure out how to get the flag to be `--no-newline` and not `--noNewline`.

I've tried setting the flags key to `'no-newline'` FWIW, no change in behavior.

@sindresorhus please advise.

sindresorhus commented:
Don't forget to also update the readme.

msabramo commented:
This now has conflicts because of the merge of #17.

Qix- commented:
That's fine, I'll fix it.

msabramo commented:
Hmmm. It looks like with the latest update `-n` works but `--no-newline` doesn't?

```
$ node cli.js -n green "Hello"; node cli.js green " world"
Hello world

$ node cli.js --no-newline green "Hello"; node cli.js green " world"
Hello
 world
```

msabramo commented:
It looks like meow is doing some fanciness where it sees the `--no` and parses it specially:

```
$ node cli.js --no-newline green "Hello"; echo
cli.flags :>>  {
  stdin: false,
  noNewline: false,
  n: false,
  demo: false,
  newline: false
}
Hello
```

msabramo commented:
This seems to work but strikes me as extremely ugly:

```diff
diff --git a/cli.js b/cli.js
index 26e428e..aea818a 100755
--- a/cli.js
+++ b/cli.js
@@ -83,7 +83,7 @@ function init(data) {

        const fn = dotProp.get(chalk, styles.join('.'));
        process.stdout.write(fn(data.replace(/\n$/, '')));
-       if (!cli.flags.noNewline) {
+       if (!cli.flags.noNewline && typeof cli.flags.newline === 'undefined') {
                process.stdout.write('\n');
        }
 }
```

```
$ node cli.js --no-newline green "Hello"; node cli.js green " world"
Hello world

$ node cli.js -n green "Hello"; node cli.js green " world"
Hello world

$ node cli.js green "Hello"; node cli.js green " world"
Hello
 world
```

Maybe @sindresorhus has ideas on how to solve this more elegantly?

msabramo commented:
Maybe slightly better version? Though still quite ugly to have to have two flags (though at least the code doesn't have to check for the absence of a flag with `typeof cli.flags.newline === 'undefined'`):

```diff
diff --git a/cli.js b/cli.js
index 26e428e..225d4b7 100755
--- a/cli.js
+++ b/cli.js
@@ -61,6 +61,10 @@ const cli = meow(`
                stdin: {
                        type: 'boolean'
                },
+               newline: {
+                       type: 'boolean',
+                       default: true
+               },
                noNewline: {
                        type: 'boolean',
                        alias: 'n'
@@ -83,7 +87,7 @@ function init(data) {

        const fn = dotProp.get(chalk, styles.join('.'));
        process.stdout.write(fn(data.replace(/\n$/, '')));
-       if (!cli.flags.noNewline) {
+       if (cli.flags.newline && !cli.flags.noNewline) {
                process.stdout.write('\n');
        }
 }
```

Qix- commented:
@msabramo see https://github.com/chalk/chalk-cli/pull/16#discussion_r707433582

msabramo commented:
Inverting the logic seems to make my first suggested code change (in https://github.com/chalk/chalk-cli/pull/16#issuecomment-918312914) slightly less ugly:

```diff
diff --git a/cli.js b/cli.js
index 26e428e..f273117 100755
--- a/cli.js
+++ b/cli.js
@@ -83,7 +83,9 @@ function init(data) {

        const fn = dotProp.get(chalk, styles.join('.'));
        process.stdout.write(fn(data.replace(/\n$/, '')));
-       if (!cli.flags.noNewline) {
+       if (cli.flags.noNewline || cli.flags.newline === false) {
+               // Skip adding a newline
+       } else {
                process.stdout.write('\n');
        }
 }
```

At least with this, it doesn't have to check if the type of a flag is `undefined`.

sindresorhus commented:
Why do you need to check `cli.flags.noNewline` at all? You already set a default value, so you can just check `cli.flags.newline`.

msabramo commented:
The check for `cli.flags.noNewline` is needed for `-n` to work. When `-n` is specified, it doesn't set `cli.flags.newline`; it only sets `cli.flags.noNewline`:

```
$ node cli.js -n green "Hello"; node cli.js green " world"
cli.flags :>>  { noNewline: true, stdin: false, demo: false }
Hello
cli.flags :>>  { stdin: false, noNewline: false, demo: false }
 world

$ node cli.js --no-newline green "Hello"; node cli.js green " world"
cli.flags :>>  { newline: false, stdin: false, noNewline: false, demo: false }
Hellocli.flags :>>  { stdin: false, noNewline: false, demo: false }
 world
```

I feel like in meow, the automatic parsing of `--no-` and the aliased flags kind of clash with each other a bit. I'm not sure what would be better though. I kind of feel like what would be cool is some way to specify an inverted option. Something kind of along the lines of the [`store_false` action in Python's argparse](https://docs.python.org/3/library/argparse.html#action) -- e.g.:

```js
		newline: {
			type: 'boolean',
			default: true,
			action: 'store_false',
			alias: 'n'
		},
```

That was what I kinda felt myself wanting in meow, but it's possible that I've missed some alternative way of accomplishing this.

msabramo commented:
Actually I guess meow doesn’t really quite need `store_false`. It already has that with the automatic parsing of `—no-`. It could maybe use a way of specifying that the **alias** specifically  is a “no flag”. 

Put another way, meow already knows that if I specify `--no-newline`, it should set `newline` to false. But if I specify `-n`, I think it’s going to set `newline` to true. It would be nice to tell meow that the alias option is inverted. 

Maybe something like:

```js 
aliasValue: false
```

?

msabramo commented:
There is another very simple thing we can do for this PR that would allow merging it and closing #13 without doing complex stuff in `chalk-cli` or `meow`.

It is to simply remove the documentation of `--no-newline` and don't support it. Just support and document `-n` (which was what was originally asked for) and don't even worry about having a long-form flag.

After all, `echo` only supports `-n` and doesn't have a long-form flag:

```
$ uname -s && man echo | grep 'trailing newline'
Darwin
     -n    Do not print the trailing newline character.  This may also be
```

```
$ docker run -it --rm ubuntu bash -c '/bin/echo --help' | grep 'trailing newline'
  -n             do not output the trailing newline
```

sindresorhus commented:
There should be a long flag. Short flags are not very user-friendly.

sindresorhus commented:
> After all, echo only supports -n and doesn't have a long-form flag:

I would not look to Unix for UX inspiration.

sindresorhus commented:
> It would be nice to tell meow that the alias option is inverted.

I agree. Would you be able to open an issue about that here https://github.com/yargs/yargs-parser ? (It's what `meow` uses for parsing)

We can use the existing code for now, but would be nice with a code comment linking to the `yarns-parser` issue.

msabramo commented:
> There should be a long flag. Short flags are not very user-friendly.

OK, agreed. Given that, another way to skirt the issue is to rename the long flag from `--no-newline` to something like `newline-omit`.

But I'm also very happy to create a yargs-parser issue, so that we don't have to hack around the issue.

msabramo commented:
> code comment linking to the yarns-parser issue

code comment linking to the yargs-parser issue

sindresorhus commented:
> Given that, another way to skirt the issue is to rename the long flag from --no-newline to something like newline-omit.

I prefer `--no-newline`. That's the convention for negated flags.

sindresorhus commented:
> But I'm also very happy to create a yargs-parser issue, so that we don't have to hack around the issue.

We'll still have to have a workaround though, as we cannot wait for yargs-parser, but it's nice to know it might eventually be fixed.

msabramo commented:
Was going to submit some PRs targeting `bare-nl` with a workaround and some other minor fixes.

Qix- commented:
Shameless plug, I wrote `arg`, I don't know how it compares to yargs though. Been a few years since I've looked. The `--no-xyz` stuff is magic to me, and it's clearly cause a bit of a headache. `arg` doesn't do any of that.

It would support this.

```javascript
import arg from 'arg';

const args = arg({
    '--no-newline': Boolean,
    '-n': '--no-newline'
});

process.stdout.write('foo');
if (!args['--no-newline']) {
    process.stdout.write('\n');
}
```

msabramo committed:
Remove extra escaping of `\n` in readme.md (#22)

There was extra escaping around `\n`, probably from copying directly
from `cli.js`. I copied from the output of `node cli.js --help`, which
gives the simpler, unescaped version.

msabramo commented:
I notice that yargs-parser supports a key called [`boolean-negation`](https://github.com/yargs/yargs-parser#boolean-negation), so *in theory* it seems like the magic can be turned off?

msabramo commented:
So I think it's possible to take advantage of the [`boolean-negation` key in yargs-parser](https://github.com/yargs/yargs-parser#boolean-negation), with a slight tweak to [meow](https://www.npmjs.com/package/meow) to allow passing custom yargs-parser options through:

```diff
diff --git a/node_modules/meow/index.js.2021-09-13-125445 b/node_modules/meow/index.js
index 6c6bb25..3f485e8 100644
--- a/node_modules/meow/index.js.2021-09-13-125445
+++ b/node_modules/meow/index.js
@@ -149,6 +149,7 @@ const meow = (helpText, options) => {

 	parserOptions.configuration = {
 		...parserOptions.configuration,
+		...options.yargsParserConfiguration,
 		'greedy-arrays': false
 	};
```

and then in `chalk-cli`:

```diff
diff --git a/cli.js b/cli.js
index 26e428e..fd88fe6 100755
--- a/cli.js
+++ b/cli.js
@@ -53,6 +53,9 @@ const cli = meow(`
          $ chalk -t '{red.bold Dungeons and Dragons {~bold.blue (with added fairies)}}'
          $ echo 'Unicorns from stdin' | chalk --stdin red bold
 `, {
+       yargsParserConfiguration: {
+               'boolean-negation': false,
+       },
        flags: {
                template: {
                        type: 'string',
```

How do we feel about that?

msabramo commented:
Workaround to make `--no-newline` work: https://github.com/chalk/chalk-cli/pull/23

I think that's not too bad, but it could be removed if we are able to fix the issue in meow (see https://github.com/chalk/chalk-cli/pull/16#issuecomment-918528286) or yargs-parser(see https://github.com/chalk/chalk-cli/pull/16#issuecomment-918473095)

msabramo committed:
Merge branch 'main' into bare-nl (#21)

msabramo committed:
Fix tests (#29)

msabramo committed:
Workaround to make --no-newline work (#23)

msabramo commented:
🎉

Qix- commented:
Thanks for the push and responsiveness @msabramo :) Great work!
